sdk/data/azcosmos/cosmos_global_endpoint_manager.go

```diff
@@ -0,0 +1,163 @@
+package azcosmos
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	azlog "github.com/Azure/azure-sdk-for-go/sdk/azcore/log"
+	azruntime "github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/log"
+)
+
+const globalEndpointManagerDefaultBackgroundRefreshInterval = 5 * time.Minute
+
+type globalEndpointManager struct {
+	client           *Client
+	preferredRegions []string
+	locationCache    *locationCache
+	refreshTicker    *time.Ticker
+	refreshDone      chan struct{}
+	sync.Mutex
+}
+
+// NewGlobalEndpointManager creates a new global endpoint manager.
+func newGlobalEndpointManager(client *Client, preferredRegions []string, refreshInterval time.Duration) (*globalEndpointManager, error) {
+	endpoint, err := url.Parse(client.endpoint)
+	if err != nil {
+		return nil, err
+	}
+	gem, err := &globalEndpointManager{
+		client:           client,
+		preferredRegions: preferredRegions,
+		locationCache:    newLocationCache(preferredRegions, *endpoint),
+	}, nil
+	if err != nil {
+		return nil, err
+	}
+	gem.startBackgroundRefresh(refreshInterval)
+
+	return gem, nil
+}
+
+// GetWriteEndpoints returns the write endpoints from the location cache.
+func (gem *globalEndpointManager) GetWriteEndpoints() ([]url.URL, error) {
+	return gem.locationCache.writeEndpoints()
+}
+
+// GetReadEndpoints returns the read endpoints from the location cache.
+func (gem *globalEndpointManager) GetReadEndpoints() ([]url.URL, error) {
+	return gem.locationCache.readEndpoints()
+}
+
+// GetAccountProperties retrieves account properties from the Cosmos DB instance.
+func (gem *globalEndpointManager) GetAccountProperties() (accountProperties, error) {
+	operationContext := pipelineRequestOptions{
+		resourceType:    resourceTypeDatabaseAccount,
+		resourceAddress: "",
+	}
+
+	path, err := generatePathForNameBased(resourceTypeDatabaseAccount, "", false)
+	if err != nil {
+		return accountProperties{}, fmt.Errorf("failed to generate path for name-based request: %v", err)
+	}
+
+	azResponse, err := gem.client.sendGetRequest(path, context.Background(), operationContext, nil, nil)
```

Comment:
For internal I/O calls, it's best practice to use a context with a timeout.  This is to prevent inadvertent hangs due to I/O failure.  Something like this (I just made up the 10 seconds, use whatever value is appropriate).
```go
ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
azResponse, err := gem.client.sendGetRequest(path, ctx, operationContext, nil, nil)
cancel()
```